### PR TITLE
header-to-metadata: support per route overrides

### PR DESCRIPTION
**Pull request by rgs1**
Signed-off-by: Raul Gutierrez Segales <rgs@pinterest.com>

### CODE REVIEW

**Comment by rgs1** [comment]
Failure seems unrelated:

```
ERROR: /source/source/extensions/filters/network/dubbo_proxy/BUILD:199:1: Linking of rule '//source/extensions/filters/network/dubbo_proxy:conn_manager_lib' failed (Exit 34): java.io.IOException: writing file failed
	at com.google.devtools.build.lib.unix.NativePosixFiles.write(Native Method)
	at com.google.devtools.build.lib.unix.UnixFileSystem$NativeFileOutputStream.write(UnixFileSystem.java:518)
	at com.google.devtools.build.lib.remote.RemoteCache$LazyFileOutputStream.write(RemoteCache.java:991)
	at com.google.common.hash.HashingOutputStream.write(HashingOutputStream.java:58)
	at java.base/java.io.FilterOutputStream.write(Unknown Source)
	at com.google.protobuf.ByteString$LiteralByteString.writeTo(ByteString.java:1320)
	at com.google.devtools.build.lib.remote.GrpcCacheClient$2.onNext(GrpcCacheClient.java:337)
	at com.google.devtools.build.lib.remote.GrpcCacheClient$2.onNext(GrpcCacheClient.java:332)
	at io.grpc.stub.ClientCalls$StreamObserverToCallListenerAdapter.onMessage(ClientCalls.java:421)
	at io.grpc.ForwardingClientCallListener.onMessage(ForwardingClientCallListener.java:33)
	at io.grpc.ForwardingClientCallListener.onMessage(ForwardingClientCallListener.java:33)
	at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1MessagesAvailable.runInContext(ClientCallImpl.java:519)
```

File: source/extensions/filters/http/header_to_metadata/header_to_metadata_filter.cc [hunk 1]
```diff
@@ -199,6 +202,34 @@ void HeaderToMetadataFilter::writeHeaderToMetadata(Http::HeaderMap& headers,
...
+const Config* HeaderToMetadataFilter::getRouteConfig() const {
+  if (!decoder_callbacks_->route() || !decoder_callbacks_->route()->routeEntry()) {
+    return nullptr;
+  }
+
+  const auto* entry = decoder_callbacks_->route()->routeEntry();
+  const auto* per_filter_config =
+      entry->virtualHost().perFilterConfig(HttpFilterNames::get().HeaderToMetadata);
+
+  return !per_filter_config ? nullptr : dynamic_cast<const Config*>(per_filter_config);
```
Comment (snowp) [inline source/extensions/filters/http/header_to_metadata/header_to_metadata_filter.cc:214]: i dont think you need this conditional, dynamic_cast of nullptr is just nullptr I believe

File: test/extensions/filters/http/header_to_metadata/header_to_metadata_filter_test.cc [hunk 2]
```diff
@@ -103,6 +110,50 @@ TEST_F(HeaderToMetadataTest, BasicRequestTest) {
...
+  Http::TestRequestHeaderMapImpl incoming_headers{{"X-VERSION", "0xdeadbeef"}};
+  std::map<std::string, std::string> expected = {{"version", "0xdeadbeef"}};
+
+  // Setup per route config.
+  envoy::extensions::filters::http::header_to_metadata::v3::Config config_proto;
+  TestUtility::loadFromYaml(request_config_yaml, config_proto);
+  Config per_route_config(config_proto, true);
+  EXPECT_CALL(decoder_callbacks_.route_->route_entry_.virtual_host_,
+              perFilterConfig(HttpFilterNames::get().HeaderToMetadata))
+      .WillOnce(testing::Return(&per_route_config));
```
Comment (snowp) [inline test/extensions/filters/http/header_to_metadata/header_to_metadata_filter_test.cc:151]: usually we do a `using testing::Return` up up to avoid having to qualify this

File: test/extensions/filters/http/header_to_metadata/header_to_metadata_filter_test.cc [hunk 3]
```diff
@@ -42,47 +72,24 @@ class HeaderToMetadataTest : public testing::Test {
 
   void initializeFilter(const std::string& yaml) {
     envoy::extensions::filters::http::header_to_metadata::v3::Config config;
-    TestUtility::loadFromYaml(yaml, config);
+    if (!yaml.empty()) {
```
Comment (snowp) [inline test/extensions/filters/http/header_to_metadata/header_to_metadata_filter_test.cc:75]: is this change necessary? mostly curious, i dont mind it

Comment (rgs1) [reply test/extensions/filters/http/header_to_metadata/header_to_metadata_filter_test.cc]: Fixed -- it wasn't necessary (the problem was I was using "" as an empty YAML config, instead of "{}").

**Comment by rgs1** [comment]
@snowp the macOS failure is transient:

```
Caused by: io.netty.channel.AbstractChannel$AnnotatedNoRouteToHostException: No route to host: remotebuildexecution.googleapis.com/2607:f8b0:4002:c02:0:0:0:5f:443
```

Can you re-kick it (assuming you have AZP super powers)?

File: source/extensions/filters/http/header_to_metadata/header_to_metadata_filter.cc [hunk 4]
```diff
@@ -16,12 +16,16 @@ namespace Extensions {
...
   response_set_ = Config::configToVector(config.response_rules(), response_rules_);
 
-  // don't allow an empty configuration
-  if (!response_set_ && !request_set_) {
+  // Note: empty configs are fine for the global config, which would be the case for enabling
+  //       the filter globally without rules and then applying them at the virtual host or
+  //       route level. At the virtual or route level, it makes no sense to have an empty
+  //       config so we throw an error.
+  if (per_route && !response_set_ && !request_set_) {
     throw EnvoyException("Must at least specify either response or request config");
```
Comment (snowp) [inline source/extensions/filters/http/header_to_metadata/header_to_metadata_filter.cc:29]: while you're in here: `Must specify at least response or request config`? maybe even specify whether its per route or global so people know what to fix

Comment (rgs1) [reply source/extensions/filters/http/header_to_metadata/header_to_metadata_filter.cc:29]: Ah yes.
